miri-script/src/commands.rs

```diff
@@ -479,6 +482,41 @@ impl Command {
         // Then test, and let caller control flags.
         // Only in root project as `cargo-miri` has no tests.
         e.test(".", &flags)?;
+
+        if coverage {
+            Self::show_coverage_report(&e)?;
+        }
+        Ok(())
+    }
+
+    fn show_coverage_report(e: &MiriEnv) -> Result<()> {
+        let profraw_files: Vec<_> = std::fs::read_dir(".")?
+        .filter_map(|r| r.ok())
+        .map(|e| e.path())
+        .filter(|p| p.extension().map(|e| e == "profraw").unwrap_or(false))
+        .map(|p| p.as_os_str().to_os_string())
+        .collect();
+
+        // Merge the profraw files
+        let profraw_files_cloned= profraw_files.iter();
+        cmd!(
+            e.sh,
+            "cargo-profdata -- merge -sparse {profraw_files_cloned...} -o merged.profdata"
+        ).quiet().run()?;
+
+        // Create the coverage report.
+        let home = std::env::var("HOME")?;
+        let ignored = format!("{home}/*|miri/target/*|target/debug/*|rust/deps/*");
```

Comment:
What exactly is this supposed to ignore? Or, what undesirable thing happens if we don't ignore anything?

Reply:
If I don't add these ignore regular expressions, the report also show files like:

- `miri/target/debug/build/chrono-tz-65fd732542149546/out/timezones.rs`
- `root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/aes-0.8.4/src/autodetect.rs `
- `rust/deps/tracing-0.1.37/src/macros.rs`

(the root path is because I run it in a container, so it is the home directory)

Additionally the coverage percentages shown in the report are smaller than the real ones, since non-miri files are included.

Ignoring these files is the simplest way I could think to fix this.

Reply:
Interesting. I suppose the report isn't supposed to look like this?
```
$ cargo-cov -- report --instr-profile=merged.profdata --object target/debug/miri -ignore-filename-regex=/home/ben/*|miri/target/*|target/debug/*|rust/deps/*
Filename                      Regions    Missed Regions     Cover   Functions  Missed Functions  Executed       Lines      Missed Lines     Cover    Branches   Missed Branches     Cover
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                               0                 0         -           0                 0         -           0                 0         -           0                 0         -
```
I do all development work from directories under `$HOME` which I think is pretty typical. So I think this regex is excluding all the profile data.

Is it possible for us to only include certain directories, instead of excluding them? We just want to include `src/` I think, which seems like a much more robust filter, if it's possible to implement.

Reply:
yeah that makes sense. Working in the container, I did my dev work under `/miri`, which was not under `$HOME` (`/root`). So I didn't think of that. Unfortunately I can't find an option to include files/directories, instead of excluding them.

We could use the [llvm_profparser](https://crates.io/crates/llvm_profparser) crate, used in tarpaulin, as i suggested in #240. It seems to provide a mapping which [can produce reports](https://docs.rs/llvm_profparser/0.7.0/llvm_profparser/coverage/coverage_mapping/struct.CoverageMapping.html#method.generate_subreport) only for the paths that we are interested in.

I went with using cargo-binutils for this PR, since it seemed like a more common crate, than tarpauling/llvm_profparser. Maybe adding llvm_profparser as a dependency for coverage reports is acceptable.

Reply:
Might be worth a feature request for cargo-binutils?

I'd rather avoid that dependency, at least for now.

Reply:
I think `--sources src/` selects the files we want. At least the coverage report it emits looks reasonable to me?

Reply:
I don't think there's any need to use `cargo-binutils`, in this case `cargo-cov` is a trivial wrapper around `llvm-cov`.

Reply:
Yes, `--sources src/` actually works. I tried using it like `--sources=src/` and it didn't work for me.

I use cargo-binutils because I saw this in the instrument coverage documentation:

```
note that the coverage mapping data generated by the Rust compiler requires LLVM version 12 or higher, and processing the raw data may require exactly the LLVM version used by the compiler. (llvm-cov --version typically shows the tool's LLVM version number, and rustc --verbose --version shows the version of LLVM used by the Rust compiler.)

    You can install compatible versions of these tools via the rustup component llvm-tools-preview. This component is the recommended path, though the specific tools available and their interface is not currently subject to Rust's usual stability guarantees. In this case, you may also find cargo-binutils useful as a wrapper around these tools.
    ```

I am not really familiar with llvm ecosystem so I don't know how easy it is to get compatible versions of the llvm tools. I used `cargo-binutils` to be more portable.

Reply:
As the cargo-binutils docs say, all that package does is exec the llvm tools from the sysroot, relying on the llvm-tools-preview component being installed. We're already doing toolchain management and finding stuff in the sysroot as part of miri-script. Can you reuse that code? Or make it reusable?

Reply:
During the `toolchain` subcommand we install the `llvm-tools` component (not the `-preview` version). I copy-pasted and modified some code from cargo-binutils and it seems to work. (do we need to do anything license-wise? I don't think so but I am not sure. Both miri and cargo-binutils have an MIT license).

After doing some more search it seems that `llvm-tools` should never have existed and the "proper" name for it should have been `llvm-tools-preview`. https://github.com/rust-lang/rust/issues/119164. However after inadvertently using the name `llvm-tools` and people started depending on it it was decided not to change the name again and break people. Indeed when I add the `llvm-tools-preview` component I see:

```
$ rustup component add llvm-tools-preview
info: downloading component 'llvm-tools'
info: installing component 'llvm-tools'
```